renter/upload.go

```diff
@@ -10,28 +10,184 @@ import (
+		// Consider a single convenience endpoint to overwrite the latest version
+		// of a file which returns an updated copy of the file object.
+		err = r.metaClient.PostFileVersion(r.Config.RenterId, existingFile.ID, newVersion)
+		if err != nil {
+			r.removeVersionBlocks(newVersion)
+			return nil, fmt.Errorf("Unable to update version metadata. Error: %s", err)
+		}
+
+		if shouldOverwrite {
+			prevVersion := &existingFile.Versions[len(existingFile.Versions)-1]
+			err = r.metaClient.DeleteFileVersion(r.Config.RenterId, existingFile.ID, prevVersion.Num)
+			if err != nil {
+				r.logger.Println("Unable to overwrite previous file version. Error:", err)
+			} else {
+				r.removeVersionBlocks(prevVersion)
+				existingFile.Versions = existingFile.Versions[:len(existingFile.Versions)-1]
+			}
+		}
+
+		// Pull down the file again to refresh the version information.
+		updatedFile, err := r.metaClient.GetFile(r.Config.RenterId, existingFile.ID)
+		if err != nil {
+			r.logger.Println("Unable to pull updated version of file. Error: %s", err)
+			existingFile.Versions = append(existingFile.Versions, *newVersion)
+			return existingFile, nil
+		}
+		*existingFile = *updatedFile
+		err = r.saveSnapshot()
+		if err != nil {
+			r.logger.Println("Error saving snapshot:", err)
+		}
+		return updatedFile, nil
+	}
+	return r.uploadFile(srcPath, finfo, destPath)
+}
+
+// Uploads a new file from srcPath to destPath. File size and destPath validation
+// should already have been performed. finfo should be the file info for the source file.
+func (r *Renter) uploadFile(srcPath string, finfo os.FileInfo, destPath string) (*core.File, error) {
+	fileId, err := genId()
+	if err != nil {
+		return nil, err
+	}
+	aesKey := make([]byte, 32)
+	_, err = rand.Reader.Read(aesKey)
+	if err != nil {
+		return nil, fmt.Errorf("Unable to create encryption key. Error: %s", err)
+	}
+	aesIV := make([]byte, aes.BlockSize)
+	_, err = rand.Reader.Read(aesIV)
+	if err != nil {
+		return nil, fmt.Errorf("Unable to read initialization vector. Error: %s", err)
+	}
+	aesKeyEncrypted, err := rsa.EncryptOAEP(sha256.New(), rand.Reader, &r.privKey.PublicKey, aesKey, nil)
+	if err != nil {
+		return nil, fmt.Errorf("Unable to encrypt aes key. Error: %s", err)
+	}
+	aesIVEncrypted, err := rsa.EncryptOAEP(sha256.New(), rand.Reader, &r.privKey.PublicKey, aesIV, nil)
+	if err != nil {
+		return nil, fmt.Errorf("Unable to encrypt aes IV. Error: %s", err)
+	}
+	version, err := r.performUpload(srcPath, finfo, aesKey, aesIV)
+	if err != nil {
+		return nil, err
+	}
+	file := &core.File{
+		ID:         fileId,
+		OwnerID:    r.Config.RenterId,
+		Name:       destPath,
+		IsDir:      false,
+		AccessList: []core.Permission{},
+		AesKey:     base64.URLEncoding.EncodeToString(aesKeyEncrypted),
+		AesIV:      base64.URLEncoding.EncodeToString(aesIVEncrypted),
+		Versions:   []core.Version{},
+	}
+	// TODO: Kincaid. Do I have to set this?
```

Comment:
No. The server will set it. If you set it, it'll just override it.

Reply:
It does that so if more than one client uploads a version at the same time, they'll get unique version numbers (it uses Mongo's atomic `$inc` operator to get the version number).